feat(docs): json blog

hasanaburayyan opened the pull request:
Blog post about Json

staycoolcall911 commented on blog/2023-04-23-json.md:
```diff
@@ -0,0 +1,112 @@
+
+While we are excited about the introduction of `Json` into the Wing type system, we are not done yet. There are still a few features we are working on, such as removing the `.get()` and `.set()` syntax for just using the `[]` operator or `.` accessor. We also have some work to do on `Json` schema validation, and some other goodies which can be found in our [language spec](https://docs.winglang.io ...
+
+As always, we welcome any feedback and would love to hear from you! Come join our
```
Let's decide what's our desired Call To Action for this blogpost and highlight a single CTA - this will allow us to later measure how effective was this blog post.

ShaiBer replied:
It's OK to have more than one CTA, if we setup our tracking properly we should still be able to measure

staycoolcall911 commented on blog/2023-04-23-json.md:
```diff
@@ -0,0 +1,112 @@
+
+That's all it takes to create a JSON object in Wing. No need to import anything—just use the `Json` keyword and get going!
+
+Now, let's create an inflight function that requires a Json object as input:
```
I think we have an issue and we currently do not support anything but `str` as input for an inflight function: https://github.com/winglang/wing/issues/2129

hasanaburayyan replied:
I updated this example to use Json.stringify() and Json.parse()

staycoolcall911 commented on blog/2023-04-23-json.md:
```diff
@@ -0,0 +1,112 @@
+
+## Mission ~~Accomplished~~ In Progress
+
+While we are excited about the introduction of `Json` into the Wing type system, we are not done yet. There are still a few features we are working on, such as removing the `.get()` and `.set()` syntax for just using the `[]` operator or `.` accessor. We also have some work to do on `Json` schema validation, and some other goodies which can be found in our [language spec](https://docs.winglang.io ...
```
I would consider telling a bit more about the struct<->Json conversions and schema validation features we have on our spec as these are the major features making Json a first class citizen in Wing and allowing this switch between strongly-typed and loosely-typed data types.

hasanaburayyan replied:
How fitting that by the time I came back to review getting this ready for publishing Im actually working on implementing this feature 😁. Ill add some details about this tomorrow when we are in New York

hasanaburayyan replied:
Section added

staycoolcall911 reviewed (commented):
Great writing @hasanaburayyan - I always enjoy reading your blogposts!
I would add a bit more about our Json/struct features which are still not implemented since they demonstrate best how we treat Json as a first class citizen

ShaiBer commented on blog/2023-04-23-json.md:
````diff
@@ -0,0 +1,112 @@
+Wing has decided to treat JSON as a first-class citizen with built-in primitives. Let's take a look at a simple example:
+
+```js
+let j = Json {
````
Got one +1 from me :)

hasanaburayyan replied:
Examples updated to show the Json keyword is optional

ShaiBer reviewed (approved):
I really love your writing @hasanaburayyan!
I have a hunch this piece can also be relevant and interesting to some other blogging platforms as well, so after we publish this, I'll try to see how to syndicate it. WDYT @Jutanium and @skorfmann ?

eladb commented:
NICE!

eladb reviewed (commented):
Let's publish this!

hasanaburayyan commented:
@eladb i forgot we didn’t 😅 I’ll try to address comments tonight so we can put it out

hasanaburayyan commented:
I updated all the out of date syntax, and added a section about Json -> Struct conversion. I think this is ready to publish if you all can give a second look ❤️

staycoolcall911 commented on blog/2023-04-23-json.md:
```diff
@@ -0,0 +1,172 @@
+
+### Converting To Structs
+
+While `Json` objects are great for interoperability, they are not the best for working with data in Wing. This is because `Json` objects are not strongly typed. However, Wing does provide a way to convert `Json` objects to structs. This conversion is done by using the `fromJson()` function, which exists on all structs created in Wing.
```
Shouldn't we also mark struct as a code-word (`struct`)? Feel free to keep it as is, of course.
```suggestion
While `Json` objects are great for interoperability, they are not the best for working with data in Wing. This is because `Json` objects are not strongly typed. However, Wing does provide a way to convert `Json` objects to `struct` instances. This conversion is done by using the static `fromJson()` method, which exists on all `struct`s created in Wing.
```

eladb replied:
I don't think it's needed. A structs is a concept in wing, like classes, loops, modules, etc.

The keyword `struct` in the language is a way to declare structs, not the other way around.

staycoolcall911 commented on blog/2023-04-23-json.md:
```diff
@@ -0,0 +1,172 @@
+
+While `Json` objects are great for interoperability, they are not the best for working with data in Wing. This is because `Json` objects are not strongly typed. However, Wing does provide a way to convert `Json` objects to structs. This conversion is done by using the `fromJson()` function, which exists on all structs created in Wing.
+
+Lets take a look at an example
```
```suggestion
Lets take a look at an example:
```

staycoolcall911 commented on blog/2023-04-23-json.md:
````diff
@@ -0,0 +1,172 @@
+```js
+bring cloud;
+
+// Struct definition of a payload object
````
We use the term "declaration" for that, while "definition" is the act of initializing a specific instance.
```suggestion
// Struct declaration of a payload object
```

staycoolcall911 commented on blog/2023-04-23-json.md:
```diff
@@ -0,0 +1,172 @@
+  content: str;
+}
+
+// inflight function that recieves json string representing a payload
```
```suggestion
// inflight function that receives json string representing a payload
```

staycoolcall911 commented on blog/2023-04-23-json.md:
```diff
@@ -0,0 +1,172 @@
+}
+
+// inflight function that recieves json string representing a payload
+new cloud.Function(inflight(msg: str) => {
```
You forgot to assign the `cloud.Function` to a variable so you could invoke it in the test below.
```suggestion
let someFunction = new cloud.Function(inflight(msg: str) => {
```

staycoolcall911 commented on blog/2023-04-23-json.md:
```diff
@@ -0,0 +1,172 @@
+
+// Test it out
+test "storing a message" {
+  let payload = Json {
```
I would delete the `Json` keyword:
```suggestion
  let payload = {
```

staycoolcall911 commented on blog/2023-04-23-json.md:
```diff
@@ -0,0 +1,172 @@
+  wow: "so easy"
+};
+
+// Optionally we can also add the Json keyword prefix
```
I saw @eladb commented on Vercel that this is redundant. I actually think it is good to keep, because later in the article it makes sense when you explain mutability and add the `MutJson` keyword.

eladb replied:
Ok fair enough

hasanaburayyan replied:
@Chriscbr I love the insight, I did not really think about it when I was updating examples. Let me swap the direction these are in

staycoolcall911 commented on blog/2023-04-23-json.md:
```diff
@@ -0,0 +1,172 @@
+
+## Mission ~~Accomplished~~ In Progress
+
+While we are excited about the introduction of `Json` into the Wing type system, we are not done yet. There are still a few features we are working on, such as removing the `.get()` and `.set()` syntax for just using the `[]` operator or `.` accessor. We also have some work to do on `Json` schema validation, and some other goodies which can be found in our [language spec](https://docs.winglang.io ...
```
Language ref link needs to be updated:
```suggestion
While we are excited about the introduction of `Json` into the Wing type system, we are not done yet. There are still a few features we are working on, such as removing the `.get()` and `.set()` syntax for just using the `[]` operator or `.` accessor. We also have some work to do on `Json` schema validation, and some other goodies which can be found in our [language spec](https://winglang.io/docs/language-reference#114-json-type).
```

hasanaburayyan replied:
amazing catch!

staycoolcall911 reviewed (approved):
This is a really great piece @hasanaburayyan! Left a few typo-fixes and enhancement suggestions which you can totally scrap. This is good to go!

ShaiBer commented on blog/2023-04-23-json.md:
```diff
@@ -0,0 +1,172 @@
+
+## Communication Is Key
+
+When designing a good distributed system, several key aspects come to mind, such as fault tolerance, scalability, security, high availability, and interoperability. In this section, we'll focus on interoperability—the ability of a system to work seamlessly with other systems. This means that communication between systems must adhere to not only standardized protocols but also data formats.
```
Don't know if it's just me, but I had to read the part with the "interoperability—the ability .." to understand it
```suggestion
When designing a good distributed system, several key aspects come to mind, such as fault tolerance, scalability, security, high availability, and interoperability. In this section, we'll focus on interoperability — the ability of a system to work seamlessly with other systems. This means that communication between systems must adhere to not only standardized protocols but also data formats.
```

Another option would be to put the explanation in parenthesis

ShaiBer commented:
Beautifully written! It is a valuable resource for existing Wing community members, but also a good piece to help us attract new ones on various subreddits (r/ProgrammingLanguages, r/programming, etc.) and HackerNews.
FYI @dannac2

eladb reviewed (approved):
Ship iiiiit
